### PR TITLE
ParamParse: Add Files at Runtime

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

Add a new runtime API to load whole files. This was currently only possible during the first initialize.

## Questions for Reviewers

- [x] does this work with MPI or do I need to call more functions?
- [ ] where shall I add a unit test for this?

## Additional background

Needed for pyAMReX and ImpactX' Python bindings.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
Yes, this works with MPI as long as all processes are calling `addfile`.

I think it's OK to not have a unit test for this.

I think it might be better to make `addfile` a static function and pass `g_table` to `addDefn`. It's `ParmParse::addfile(..)` vs. `ParmParse().addfile(..)`.  The later could be misleading because the user might have the question what happens after the temporary `ParmParse()` is gone.

**Comment by WeiqunZhang** [comment]
Yes, this works with MPI as long as all processes are calling `addfile`.

I think it's OK to not have a unit test for this.

I think it might be better to make `addfile` a static function and pass `g_table` to `addDefn`. It's `ParmParse::addfile(..)` vs. `ParmParse().addfile(..)`.  The later could be misleading because the user might have the question what happens after the temporary `ParmParse()` is gone.

**Comment by ax3l** [comment]
Thanks!

For MPI, I would prefer if we read from the IO processor and bcast. I have not yet found the API for that, can you pls point me there?

Static: I see. I thought this is added to some static/global state anyway, at least that how it appeared. Let's rewrite if that's clearer, thx.

**Comment by WeiqunZhang** [comment]
Yes, `addDefn` will have the I/O process read and bcast.
https://github.com/AMReX-Codes/amrex/blob/6f72de283c38ef3e38ed7c5e8760a176434e6be3/Src/Base/AMReX_ParmParse.cpp#L494
https://github.com/AMReX-Codes/amrex/blob/6f72de283c38ef3e38ed7c5e8760a176434e6be3/Src/Base/AMReX_ParallelDescriptor.cpp#L1393

**Comment by ax3l** [comment]
Ah perfect, so with regards to MPI there is nothing that needs to be changed and the only thing is that the new API call is a collective call. Great!

File: Src/Base/AMReX_ParmParse.H [hunk 1]
```diff
@@ -552,6 +558,13 @@ public:
     //! Add a key 'name'with value 'ref' to the end of the PP table.
     void add (const char* name,
               const std::string&  val);
+
+    //! keyword for files to load
+    static std::string FileKeyword;
+
+    //! Add keys and values from a file to the end of the PP table.
+    static void addfile (std::string const filename, ParmParse::Table& tab);
```
Comment (ax3l) [inline Src/Base/AMReX_ParmParse.H]: @WeiqunZhang oh, how can I pass a table object here? Our `table()` method is only const, so this does not work:
```C++
amrex::ParmParse pp;
pp.addfile(filename, pp.table());
```

```
error: binding reference of type ‘amrex::ParmParse::Table&’ {aka ‘std::__cxx11::list<amrex::ParmParse::PP_entry>&’} to ‘const Table’ {aka ‘const std::__cxx11::list<amrex::ParmParse::PP_entry>’} discards qualifiers
   44 |                 pp.addfile(filename, pp.table());
      |                                      ~~~~~~~~^~
```

Comment (ax3l) [reply Src/Base/AMReX_ParmParse.H]: oh, `g_table` internally...

**Comment by ax3l** [comment]
@WeiqunZhang I converted the method to be static :)

**Review by WeiqunZhang — APPROVED** [review]

**Pull request merged**